libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -185,31 +186,29 @@ proc getGossipPeers(g: GossipSub): Table[string, ControlMessage] {.gcsafe.} =
     let fanout: HashSet[string] = g.fanout.getOrDefault(topic)
 
     let gossipPeers = mesh + fanout
+    var extraPeers = g.gossipsub # copy it!
     let mids = g.mcache.window(topic)
     if mids.len > 0:
       let ihave = ControlIHave(topicID: topic,
                                messageIDs: toSeq(mids))
 
-      if topic notin g.gossipsub:
+      if topic notin extraPeers:
         trace "topic not in gossip array, skipping", topicID = topic
         continue
 
       while result.len < GossipSubD:
-        if g.gossipsub.getOrDefault(topic).len == 0:
+        if extraPeers.getOrDefault(topic).len == 0:
           trace "no peers for topic, skipping", topicID = topic
           break
 
-        let id = toSeq(g.gossipsub.getOrDefault(topic)).sample()
-        if id in g.gossipsub.getOrDefault(topic):
-          g.gossipsub[topic].excl(id)
+        let id = toSeq(extraPeers.getOrDefault(topic)).sample()
+        if id in extraPeers.getOrDefault(topic):
+          extraPeers[topic].excl(id)
```

Comment:
This was the bug, we should not be excluding the peers from the gossip table and the `extraPeers` table doesn't make sense here. The algorithm for gossip peers is actually this https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md#gossip-emission, you might notice that the `mcache.shift()` is absent, that is because it is done at the end of the heartbeat cycle:
 https://github.com/status-im/nim-libp2p/blob/f49e59cb4e86b8342618a18062c88ef5d15f11aa/libp2p/protocols/pubsub/gossipsub.nim#L233